First draft on the dedicated JSON page (#2088)
RediSearch/RediSearch: docs/Indexing_JSON.md

# Indexing JSON documents

> As a secondary index, RediSearch aims to index data structures available in Redis.

Reviewer: ```suggestion
In addition to indexing Redis hashes, RediSearch also indexes JSON. To index JSON, you must use the RedisJSON module.
```

> In its first implementations (1.x, 2.0) RediSearch supported only hashes.

Reviewer: ```suggestion

```

> Associated with RediJSON (version 2.0 or more recent), RediSearch is able to index JSON documents.

Reviewer: ```suggestion

```

## Prerequisite

> What do you need to start indexing JSON documents ?

Reviewer: ```suggestion
What do you need to start indexing JSON documents?
```

> - Redis 6.0x (or more recent)

Reviewer: ```suggestion
- Redis 6.x or later
```

> - The module RediSearch 2.2 (or more recent)

Reviewer: ```suggestion
- RediSearch 2.2 or later
```

> - The module RediJSON 2.0 (or more recent)

Reviewer: ```suggestion
- RediJSON 2.0 or later
```

> ## Creating an index for JSON documents indexing

Reviewer: ```suggestion
## How to index JSON documents
```

The [FT.CREATE](Commands.md#FT.CREATE) command has evolved to support the new data type.
We can now specify `ON JSON` to inform RediSearch that we want to index JSON documents.

Then, on the `SCHEMA` part, you can provide JSON Path expressions.
The result of each `JSON Path` expression is indexed and associated with a logical name (`attribute`).
This attribute (previously called `field`) is used in the query part.

This is the minimal syntax we have to use to be able to index a JSON document:

    FT.CREATE {index_name} ON JSON SCHEMA {json_path} AS {attribute} {type}

Here is a concrete example:

    FT.CREATE myIdx ON JSON SCHEMA $.user.name AS name TEXT $.user.tag AS tag TAG

## Adding JSON document to the index

As soon as the index is created, any pre-existing JSON document,
or any new JSON document added or modified is automatically indexed.

We are free to use any writing command from the RedisJSON module (`JSON.SET`, `JSON.ARRAPPEND`, etc.).

In our example we are going to use the following JSON document:

```JSON
{
  "user": {
    "name": "John Smith",
    "tag": ["foo","bar"],
    "hp": "1000",
    "dmg": "150"
  }
}
```

We can use `JSON.SET` to store in our database:

>     JSON.SET myDoc . '{"user":{"name":"John Smith","tag":["foo","bar"],"hp":1000, "dmg":150}}'

Reviewer 2: ```suggestion
    JSON.SET myDoc $ '{"user":{"name":"John Smith","tag":["foo","bar"],"hp":1000, "dmg":150}}'
```

Following our example, let's find our user called `John`:

```
FT.SEARCH myIdx '@name:(John)'
1) (integer) 1
2) "myDoc"
3) 1) "$"
   2) "{\"user\":{\"name\":\"John Smith\",\"tag\":[\"foo\",\"bar\"],\"hp\":1000,\"dmg\":150}}"
```

## Field projection

We just saw that, by default, `FT.SEARCH` returns the whole document.

We can return only specific attribute (here `name`):

```
FT.SEARCH myIdx '@name:(John)' RETURN 1 name
1) (integer) 1
2) "myDoc"
3) 1) "name"
   2) "\"John Smith\""
```

### Projecting using JSON Path expressions

The `RETURN` parameter also accepts `JSON Path expression` which let us extract any part of the JSON document.

In this example, we return the result of the JSON Path expression `$.user.hp`.

```
FT.SEARCH myIdx '@name:(John)' RETURN 1 $.user.hp
1) (integer) 1
2) "myDoc"
3) 1) "$.user.hp"
   2) "1000"
```

Note that the property name is the JSON expression itself: `3) 1) "$.user.hp"`

Using the `AS` option, it is also possible to change then name of the returned property.

```
FT.SEARCH myIdx '@name:(John)' RETURN 3 $.user.hp AS hitpoints
1) (integer) 1
2) "myDoc"
3) 1) "hitpoints"
   2) "1000"
```

### Highlighting

> On JSON, we can highlight any attribute as soon as he is indexed using the TEXT type.

Reviewer 2: ```suggestion
On JSON, we can highlight any attribute as soon as it is indexed using the TEXT type.
```
